methcomp/regressor.py

```diff
@@ -0,0 +1,498 @@
+        standard errors (and confidence intervals).
+    result : Dict[str, Any]
+        Regression result with
+            `slope`, `intercept`, `sx`, and `sy`
+        if bootstrap > 0 each of these contain
+            `value`, `ci_low`, `ci_high`, and `SE`
+
+    sdr : float
+        The assumed known standard deviations.
+    vr : float
+        The assumed known ratio of the (residual) variance of the ys relative
+        to that of the xs.
+
+
+    References
+    ----------
+    .. [Koopmans_1937] Koopmans, T. C. (1937).
+                       "Linear regression analysis of economic time series."
+                       DeErven F. Bohn, Haarlem, Netherlands.
+    .. [Deming_1943] Deming, W. E. (1943).
+                     "Statistical adjustment of data."
+                     Wiley, NY (Dover Publications edition, 1985).
+    """
+
+    def __init__(
+        self,
+        method1: Union[List[float], np.ndarray],
+        method2: Union[List[float], np.ndarray],
+        CI: float = 0.95,
+        vr: float = None,
+        sdr: float = None,
+        bootstrap: int = 1000,
+    ):
+        """Construct a Deming Regressor
+
+        Parameters
+        ----------
+        method1 : Union[List[float], np.ndarray]
+            Values for method 1
+        method2 : Union[List[float], np.ndarray]
+            Values for method 2
+        CI : float, optional
+            The confidence interval employed in regression line
+            [default=0.95]
+        vr : float, optional
+            The assumed known ratio of the (residual) variance of the ys
+            relative to that of the xs.
+            [default=1]
+        sdr : float, optional
+            The assumed known standard deviations. Parameter vr takes
+            precedence if both are given.
+            [default=1]
+        bootstrap : int, optional
+            Amount of bootstrap estimates that should be performed to acquire
+            standard errors (and confidence intervals). If None, no bootstrap
+            is performed.
+            [default=1000]
+        """
+        self.vr = vr
+        self.sdr = sdr
+        self.bootstrap = bootstrap
+        super().__init__(method1, method2, CI)
+
+    def _check_params(self):
+        """Check validity of parameters
+
+        Raises
+        ------
+        ValueError
+            If method values are of different shape or CI outside of range 0,1
+        """
+        if self.vr is not None < 0:
+            raise ValueError("vr parameter must be positive or None")
+        if self.sdr is not None < 0:
+            raise ValueError("sdr parameter must be positive or None")
+        if self.bootstrap is not None and self.bootstrap <= 0:
+            raise ValueError("bootstrap parameter must be postivie or None")
+
+    def _calculate_impl(self):
+        """Calculate regression parameters."""
+
+        def deming(n: int, x: np.ndarray, y: np.ndarray, lamb: float) -> np.ndarray:
```

Comment:
It looks a tiny bit odd to have an internal `deming()` function now within the `_calculate_impl` but I understand its the most cleanest for now - so we can just keep it.

Reply:
I will rename it to `_calc_deming`, it has to be a function as it is called in 2 places, but it certainly doesn't have to be called `deming`